Python/Product/PythonTools/PythonTools/Editor/PythonTextBufferInfo.cs

```diff
@@ -739,4 +744,157 @@ private Lazy<Tokenizer> GetTokenizerLazy()
 
         #endregion
     }
+
+    static class PythonTextBufferInfoExtensions {
+        public static PythonTextBufferInfo TryGetInfo(this ITextBuffer buffer) => PythonTextBufferInfo.TryGetForBuffer(buffer);
+
+        public static AnalysisEntry TryGetAnalysisEntry(this ITextBuffer buffer) => PythonTextBufferInfo.TryGetForBuffer(buffer)?.AnalysisEntry;
+
+        public static Task<AnalysisEntry> GetAnalysisEntryAsync(this ITextBuffer buffer, PythonEditorServices services = null, CancellationToken cancellationToken = default(CancellationToken)) {
+            var bi = services == null ? PythonTextBufferInfo.TryGetForBuffer(buffer) : services.GetBufferInfo(buffer);
+            if (bi != null) {
+                return bi.GetAnalysisEntryAsync(cancellationToken);
+            }
+            return Task.FromResult<AnalysisEntry>(null);
+        }
+
+        public static AnalysisEntry TryGetAnalysisEntry(this ITextView view, IServiceProvider site) {
+            var entry = view.TextBuffer.TryGetAnalysisEntry();
+            if (entry != null) {
+                return entry;
+            }
+
+            var diffViewer = site.GetComponentModel().GetService<IWpfDifferenceViewerFactoryService>();
+            var viewer = diffViewer?.TryGetViewerForTextView(view);
+            if (viewer != null) {
+                entry = viewer.DifferenceBuffer.RightBuffer.TryGetAnalysisEntry() ??
+                    viewer.DifferenceBuffer.LeftBuffer.TryGetAnalysisEntry();
+                if (entry != null) {
+                    return entry;
+                }
+            }
+
+            return null;
+        }
+
+        public static async Task<ProjectAnalyzer> FindAnalyzerAsync(this IServiceProvider site, ITextView view) {
+            ProjectAnalyzer analyzer;
+
+            var bi = view.TextBuffer.TryGetInfo();
+            if (bi != null && (analyzer = await FindAnalyzerAsync(site, bi)) != null) {
+                return analyzer;
+            }
+
+            var diffViewer = site.GetComponentModel().GetService<IWpfDifferenceViewerFactoryService>();
```

Comment:
Should we check for UI thread?

Reply:
Because this is accessing via ITextView, I think it's okay to require the function be called from the UI thread. I'll add an assert.